addons/delivery/models/delivery_carrier.py

```diff
@@ -123,3 +145,129 @@ def cancel_shipment(self, pickings):
+
+    @api.multi
+    def carrier_get(self, contact):
+        self.ensure_one()
+        country_ids = self.country_ids.ids
+        state_ids = self.state_ids.ids
+        if country_ids and not contact.country_id.id in country_ids:
+            return False
+        if state_ids and not contact.state_id.id in state_ids:
+            return False
+        if self.zip_from and (contact.zip or '') < self.zip_from:
+            return False
+        if self.zip_to and (contact.zip or '') > self.zip_to:
+            return False
+        return self
+
+    @api.multi
+    def create_price_rules(self):
+        PriceRule = self.env['delivery.price.rule']
+        for record in self:
+            # if using advanced pricing per destination: do not change
+            if record.delivery_type == 'base_on_rule':
+                continue
+
+            # not using advanced pricing per destination: override lines
+            if record.delivery_type == 'base_on_rule' and not (record.fixed_price is not False or record.free_if_more_than):
+                record.price_rule_ids.unlink()
+
+            # Check that float, else 0.0 is False
+            if not (record.fixed_price is not False or record.free_if_more_than):
+                continue
+
+            if record.delivery_type == 'fixed':
+                PriceRule.search([('carrier_id', '=', record.id)]).unlink()
+
+            #create the delivery price rules
+            if record.free_if_more_than:
+                line_data = {
+                    'carrier_id': record.id,
+                    'name': _('Free if more than %.2f') % record.amount,
+                    'variable': 'price',
+                    'operator': '>=',
+                    'max_value': record.amount,
+                    'standard_price': 0.0,
+                    'list_base_price': 0.0,
+                }
+                PriceRule.create(line_data)
+            if record.fixed_price is not False:
+                line_data = {
+                    'carrier_id': record.id,
+                    'name': _('Default price'),
+                    'variable': 'price',
+                    'operator': '>=',
+                    'max_value': 0.0,
+                    'standard_price': record.fixed_price,
+                    'list_base_price': record.fixed_price,
+                }
+                PriceRule.create(line_data)
+        return True
+
+    def filter_state_country(self, vals):
+        if vals.get('country_ids'):
+            country_ids = [country['id'] for country in self.resolve_2many_commands('country_ids', vals.get('country_ids'))]
+            state_ids = [state['id'] for state in self.resolve_2many_commands('state_ids', vals.get('state_ids'))]
+            vals['state_ids'] = [(6, 0, self.env['res.country.state'].browse(state_ids).filtered(lambda x: x.country_id.id in country_ids).ids)]
+        return vals
+
+    @api.model
+    def create(self, vals):
+        vals = self.filter_state_country(vals)
```

Comment:
after call super create method, can we delete link between state and delivery method if state is not in specified country list ?